sa/storage-authority.go

```diff
@@ -336,181 +418,143 @@ func (ssa *SQLStorageAuthority) MarkCertificateRevoked(serial string, ocspRespon
+	pending_authz := &Pending_auth{Authorization: core.Authorization{ID: id}}
+	err = ssa.dbMap.Insert(pending_authz)
 	return
 }
 
 func (ssa *SQLStorageAuthority) UpdatePendingAuthorization(authz core.Authorization) (err error) {
-	tx, err := ssa.db.Begin()
-	if err != nil {
-		return
-	}
-
 	if !statusIsPending(authz.Status) {
 		err = errors.New("Use Finalize() to update to a final status")
-		tx.Rollback()
 		return
 	}
 
-	if existingFinal(tx, authz.ID) > 0 {
+	if ssa.existingFinal(authz.ID) {
 		err = errors.New("Cannot update a final authorization")
-		tx.Rollback()
 		return
 	}
 
-	if existingPending(tx, authz.ID) != 1 {
+	if !ssa.existingPending(authz.ID) {
 		err = errors.New("Requested authorization not found " + authz.ID)
-		tx.Rollback()
 		return
 	}
 
-	jsonAuthz, err := json.Marshal(authz)
+	authObj, err := ssa.dbMap.Get(Pending_auth{}, authz.ID)
 	if err != nil {
-		tx.Rollback()
 		return
 	}
-
-	_, err = tx.Exec("UPDATE pending_authz SET value = ? WHERE id = ?;", jsonAuthz, authz.ID)
-	if err != nil {
-		tx.Rollback()
-		return
-	}
-
-	err = tx.Commit()
+	auth := authObj.(*Pending_auth)
+	auth.Authorization = authz
+	_, err = ssa.dbMap.Update(auth)
 	return
 }
 
 func (ssa *SQLStorageAuthority) FinalizeAuthorization(authz core.Authorization) (err error) {
-	tx, err := ssa.db.Begin()
-	if err != nil {
-		return
-	}
-
 	// Check that a pending authz exists
-	if existingPending(tx, authz.ID) != 1 {
+	if !ssa.existingPending(authz.ID) {
 		err = errors.New("Cannot finalize a authorization that is not pending")
-		tx.Rollback()
 		return
 	}
 	if statusIsPending(authz.Status) {
 		err = errors.New("Cannot finalize to a non-final status")
-		tx.Rollback()
 		return
 	}
 
 	// Manually set the index, to avoid AUTOINCREMENT issues
 	var sequence int64
-	var scanTarget sql.NullInt64
-	err = tx.QueryRow("SELECT max(sequence) FROM authz").Scan(&scanTarget)
+	sequenceObj, err := ssa.dbMap.SelectNullInt("SELECT max(sequence) FROM authz")
 	switch {
-	case !scanTarget.Valid:
+	case !sequenceObj.Valid:
 		sequence = 0
 	case err != nil:
-		tx.Rollback()
 		return
 	default:
-		sequence += scanTarget.Int64 + 1
+		sequence += sequenceObj.Int64 + 1
 	}
 
 	jsonAuthz, err := json.Marshal(authz)
 	if err != nil {
-		tx.Rollback()
 		return
 	}
+	// ???: is this still needed? ^+v
 	digest := core.Fingerprint256(jsonAuthz)
 
-	// Add to final table and delete from pending
+	auth := &Auth{sequence, digest, authz}
+	authObj, err := ssa.dbMap.Get(Pending_auth{}, authz.ID)
 	if err != nil {
-		tx.Rollback()
 		return
 	}
-	_, err = tx.Exec("INSERT INTO authz (sequence, id, digest, value) VALUES (?, ?, ?, ?);", sequence, authz.ID, digest, jsonAuthz)
+	oldAuth := authObj.(*Pending_auth)
+
+	err = ssa.dbMap.Insert(auth)
 	if err != nil {
-		tx.Rollback()
 		return
 	}
-	_, err = tx.Exec("DELETE FROM pending_authz WHERE id = ?;", authz.ID)
-	if err != nil {
-		tx.Rollback()
-		return
-	}
-	err = tx.Commit()
+
+	_, err = ssa.dbMap.Delete(oldAuth)
```

Comment:
It seems like the insert and delete here should definitely stay wrapped in a transaction, or we can get into an inconsistent state.

I think in the long run we may want to eliminate the separate tables for pending / final authorizations, but that's a change for another day.